src/glum/_glm.py

```diff
@@ -533,37 +538,41 @@ def setup_p1(
 
     # P1 and P2 are now for sure copies
     P1 = alpha * l1_ratio * P1
-    return cast(np.ndarray, P1).astype(_dtype)
+    return cast(np.ndarray, P1).astype(dtype)
 
 
 def setup_p2(
-    P2: Union[str, np.ndarray, sparse.spmatrix],
+    P2: Optional[Union[str, np.ndarray, sparse.spmatrix]],
```

Comment:
In a month or so, when we drop support for python 3.9 we will be able to use the much cleaner (IMO) `type | None`.